Manifold/Error.swift

```diff
@@ -46,31 +75,16 @@ public enum Error: Printable, StringLiteralConvertible {
 
 /// Constructs a composite error.
 public func + (left: Error, right: Error) -> Error {
-	switch (left, right) {
-	case (.Leaf, .Leaf):
-		return .Branch([ left, right ])
-
-	case let (.Leaf, .Branch(errors)):
-		return .Branch([ left ] + errors)
-
-	case let (.Branch(errors), .Leaf):
-		return .Branch(errors + [ right ])
-
-	case let (.Branch(e1), .Branch(e2)):
-		return .Branch(e1 + e2)
-	}
+	return Error.Branch(left.errors + right.errors)
 }
 
 
-/// The logical and of two Either<Error>s.
-public func && <A, B> (a: Either<Error, A>, b: Either<Error, B>) -> Either<Error, (A, B)> {
-	return a.either(
-		{ (a: Error) in b.either(
-			{ Either.left(a + $0) },
-			const(Either.left(a))) },
-		{ (a: A) in b.either(
-			Either.left,
-			{ Either.right(a, $0) }) })
+/// Computes the conjunction of two `Either`s.
+public func &&& <T, U> (a: Either<Error, T>, b: Either<Error, U>) -> Either<Error, (T, U)> {
+	let right = (a.right &&& b.right).map(Either<Error, (T, U)>.right)
+	let lefts = (a.left &&& b.left).map(+).map(Either<Error, (T, U)>.left)
+	let left = (a.left.map(Either<Error, (T, U)>.left) ||| b.left.map(Either<Error, (T, U)>.left))?.either(id, id)
+	return (right ?? lefts ?? left)!
```

Comment:
This runs into a pretty terrible bug still extant in Swift 1.2b1 where the `left` and `right` properties will _nondeterministically_ each return `nil` for the same `Either`. Editing `Either` to use `switch` instead of `either` in the implementations of those properties appears to work around the issue.

I’m hopeful it will be fixed in 1.2b2, so I haven’t yet committed that patch to Either.

cf rdar://19673469